Parser/asdl_c.py

```diff
@@ -1132,13 +1132,93 @@ def visitModule(self, mod):
     return result;
 }
 
+/* copy.replace() support (shallow copy) */
+static PyObject *
+ast_type_replace(PyObject *self, PyObject *args, PyObject *kwargs) {
+    if (!_PyArg_NoPositional("__replace__", args)) {
+        return NULL;
+    }
+
+    struct ast_state *state = get_ast_state();
+    if (state == NULL) {
+        return NULL;
+    }
+
+    PyObject *result = NULL;
+    // known AST class fields
+    PyObject *fields = NULL;
+    // current instance dictionary
+    PyObject *__dict__ = NULL;
+    // constructor positional and keyword arguments
+    PyObject *empty_tuple = NULL, *payload = NULL;
+    // for iterating over 'fields' and 'kwargs'
+    PyObject *key = NULL, *value = NULL;
+
+    PyTypeObject *type = Py_TYPE(self);
+    if (PyObject_GetOptionalAttr((PyObject *)type, state->_fields, &fields) < 0) {
+        goto cleanup;
+    }
+    if (PyObject_GetOptionalAttr(self, state->__dict__, &__dict__) < 0) {
+        goto cleanup;
+    }
+    if (!(empty_tuple = PyTuple_New(0))) {
+        goto cleanup;
+    }
+    if (!(payload = PyDict_New())) {
+        goto cleanup;
+    }
+    if (fields) {
+        Py_ssize_t numfields = PySequence_Size(fields);
+        if (numfields == -1) {
+            goto cleanup;
+        }
+        for (Py_ssize_t i = 0; i < numfields; i++) {
+            if (!(key = PySequence_GetItem(fields, i))) {
+                goto cleanup;
+            }
+            if (PyDict_GetItemRef(__dict__, key, &value) < 0) {
+                goto cleanup;
+            }
+            if (!value) {
+                break;
+            }
+            if (PyDict_SetItem(payload, key, value) < 0) {
+                goto cleanup;
+            }
+        }
+    }
+    if (kwargs && PyDict_Update(payload, kwargs) < 0) {
+        goto cleanup;
+    }
+    if (!(result = type->tp_new(type, empty_tuple, payload))) {
+        Py_CLEAR(result);
+        goto cleanup;
+    }
+    if (ast_type_init(result, empty_tuple, payload) < 0) {
+        Py_CLEAR(result);
+        goto cleanup;
+    }
```

Comment:
The same suggestion about assignment in expression.

And why not simply use `PyObject_Call((PyObject *)type, empty_tuple, payload)`?

Reply:
Err... I wasn't sure which one is the most efficient way. By calling them separately I thought it would reduce the number of checks but if that's the standard way to do I'll do it (thanks for the advice!)

Reply:
This code does not use `__new__` and `__init__` if they are overridden in a subclass. Unless you have a reason for such behavior, it is better to just call `type`.